python_modules/dagster/dagster_tests/storage_tests/test_staging_io_manager.py

```diff
@@ -0,0 +1,262 @@
+                )
+            )
+        ]
+
+
+def test_basic_bound_runner_usage():
+    now_time_plus_15 = get_now_time_plus_N(15)
+    with instance_for_test() as instance:
+        runner = BoundRunner(
+            Definitions(
+                assets=[now_time],
+                resources={"io_manager": fs_io_manager},
+            ),
+            instance,
+        )
+
+        assert runner.materialize_all_assets().success
+        assert isinstance(runner.load_asset_value("now_time"), int)
+
+
+def get_env_entry(event_log_entry: EventLogEntry) -> Optional[str]:
+    asset_mat = get_asset_materialization_from_event_log_entry(event_log_entry)
+    if not asset_mat:
+        return None
+
+    return get_text_metadata(asset_mat.metadata_entries, label="env")
+
+
+def test_write_staging_label():
+    with DagsterInstance.ephemeral() as instance:
+        staging_runner = BoundRunner(
+            Definitions(
+                assets=[now_time],
+                resources={
+                    "io_manager": DualEnvironmentIOManager(
+                        prod_io_manager=InMemoryIOManager(),
+                        staging_io_manager=InMemoryIOManager(),
+                    )
+                },
+            ),
+            instance,
+        )
+
+        assert staging_runner.materialize_all_assets().success
+
+        all_mat_log_records = staging_runner.get_all_asset_materialization_event_records("now_time")
+        assert all_mat_log_records
+
+        assert len(all_mat_log_records) == 1
+        asset_mat = get_asset_materialization_from_event_log_entry(
+            all_mat_log_records[0].event_log_entry
+        )
+
+        assert asset_mat
+
+        assert get_text_metadata(asset_mat.metadata_entries, label="env") == "staging"
+
+
+def test_write_name_staging_label():
+    with DagsterInstance.ephemeral() as instance:
+        staging_runner = BoundRunner(
+            Definitions(
+                assets=[now_time],
+                resources={
+                    "io_manager": DualEnvironmentIOManager(
+                        prod_io_manager=InMemoryIOManager(),
+                        staging_io_manager=InMemoryIOManager(),
+                        env_key="another_env",
+                    ),
+                },
+            ),
+            instance,
+        )
+
+        assert staging_runner.materialize_all_assets()
+
+        all_mat_log_records = staging_runner.get_all_asset_materialization_event_records("now_time")
+        assert all_mat_log_records
+
+        assert len(all_mat_log_records) == 1
+        asset_mat = get_asset_materialization_from_event_log_entry(
+            all_mat_log_records[0].event_log_entry
+        )
+
+        assert asset_mat
+
+        assert get_text_metadata(asset_mat.metadata_entries, label="another_env") == "staging"
+
+
+@asset
+def now_time_plus_20_after_plus_N(now_time_plus_N: int) -> int:
+    return now_time_plus_N + 20
+
+
+def test_basic_workflow():
```

Comment:
This is the test that demonstrates the workflow where one is iterating on asset in the middle of an asset graph.